### PR TITLE
[openwebnet] adding support for central unit

### PR DESCRIPTION
**Pull request by aconte80**
Objective:

Enhance OpenWebNet binding to support thermoregulation with 4/99 zone central unit, e.g. BTicino 3550 model.

**Changelog**

- manual setting of a zone to T temperature (with persistence in central unit)
- manual setting of central unit to T temperature (all zones)
- read Battery status
- read Remote Control status
- set the central unit in OFF mode (all zones)
- set central unit in thermal protection (all zones)
- set central unit in antifreeze mode (all zones)
- weekly program activation command (all zones)
- scenario activation command (all zones)

### CODE REVIEW

**Comment by lolodomo** [comment]
What's the status of this PR? Is it still work in progress like the title suggests?

**Comment by aconte80** [comment]
@lolodomo I've removed the wip tag

**Comment by aconte80** [comment]
This PR will also solve https://github.com/openhab/openhab-addons/issues/12298

**Review by lolodomo — CHANGES_REQUESTED** [review]

File: bundles/org.openhab.binding.openwebnet/src/main/resources/OH-INF/thing/BusThermoCentralUnit.xml [hunk 1]
```diff
@@ -0,0 +1,35 @@
...
+		</channels>
+
+		<properties>
+			<property name="vendor">BTicino/Legrand</property>
+			<property name="model">BTicino Central Unit 3550 (99 zones) or BTI-L/NT/HC/4695 (4 zones) </property>
+			<property name="ownDeviceType">430</property>
+		</properties>
+
+		<representation-property>ownId</representation-property>
+	</thing-type>
```
[the remark this answers is not in the document]
Comment (aconte80) [reply bundles/org.openhab.binding.openwebnet/src/main/resources/OH-INF/thing/BusThermoCentralUnit.xml:34]: The central unit has a "where" but the only allowed value is 0. 
This is why I didn't add a configuration parameter like in ThermoSensor or ThermoZone.

File: bundles/org.openhab.binding.openwebnet/src/main/resources/OH-INF/thing/channels.xml [hunk 2]
```diff
@@ -170,6 +170,103 @@
...
+			<options>
+				<option value="1">Program 1</option>
+				<option value="2">Program 2</option>
+				<option value="3">Program 3</option>
+			</options>
+		</state>
+	</channel-type>
+
+	<channel-type id="scenarioProgramCentralUnit">
+		<item-type>String</item-type>
```
[the remark this answers is not in the document]
Comment (aconte80) [reply bundles/org.openhab.binding.openwebnet/src/main/resources/OH-INF/thing/channels.xml:221]: you're right, it was a typo

File: bundles/org.openhab.binding.openwebnet/README.md [hunk 3]
```diff
@@ -173,13 +176,19 @@ Systems with Central Units (4 or 99 zones) are not fully supported yet.
...
+| `localOffset`          | `bus_thermo_zone`                    | String             | The zone local offset status: `OFF`, `PROTECTION`, `MINUS_3`, `MINUS_2` , `MINUS_1`, `NORMAL`, `PLUS_1`, `PLUS_2`, `PLUS_3`  | R | Y |
+| `remoteControl`          | `bus_thermo_cu`                    | String             | The Central Unit Remote Control status: `ENABLED`, `DISABLED`  | R | Y |
+| `batteryStatus`          | `bus_thermo_cu`                    | String             | The Central Unit Battery status: `OK`, `KO`  | R | Y |
+| `weeklyProgram`                       | `bus_thermo_cu`                    | String             | The program number (`1`, `2`, `3`) when Central Unit mode is `WEEKLY`  | R/W        | N        |
+| `scenarioProgram`                       | `bus_thermo_cu`                    | String             | The program number (`1`, `2`, .. ,  `16`) when Central Unit mode is `SCENARIO` | R/W        | N        |
```
Comment (lolodomo) [inline bundles/org.openhab.binding.openwebnet/README.md:190]: Type is now Number

File: bundles/org.openhab.binding.openwebnet/README.md [hunk 4]
```diff
@@ -255,17 +277,25 @@ Bridge openwebnet:zb_gateway:myZBgateway  [ serialPort="COM3" ] {
...
 Number:Power        iCENTRAL_Tb           "Power [%.0f %unit%]" { channel="openwebnet:bus_energy_meter:mybridge:CENTRAL_Tb:power" }
 
+// 99 zones central unit 
+Group   gCentralUnit                "Central Unit"                
+Number:Temperature iCU_3550_manualset "Temperature"       (gCentralUnit) { channel="openwebnet:bus_thermo_cu:mybridge:CU_3550:setpointTemperature", ga="thermostatTemperatureSetpoint" }
+String iCU_3550_remote    "Remote Control"    (gCentralUnit) { channel="openwebnet:bus_thermo_cu:mybridge:CU_3550:remoteControl" }
+String iCU_3550_battery   "Battery Status"    (gCentralUnit) { channel="openwebnet:bus_thermo_cu:mybridge:CU_3550:batteryStatus" }
+String iCU_3550_mode      "Mode"              (gCentralUnit) { channel="openwebnet:bus_thermo_cu:mybridge:CU_3550:mode" }
+String iCU_3550_wpn       "Weekly Program"    (gCentralUnit) { channel="openwebnet:bus_thermo_cu:mybridge:CU_3550:weeklyProgram" } 
+String iCU_3550_spn       "Scenario Program" (gCentralUnit) { channel="openwebnet:bus_thermo_cu:mybridge:CU_3550:scenarioProgram" }
```
Comment (lolodomo) [inline bundles/org.openhab.binding.openwebnet/README.md:298]: Type is now Number

File: bundles/org.openhab.binding.openwebnet/README.md [hunk 5]
```diff
@@ -255,17 +277,25 @@ Bridge openwebnet:zb_gateway:myZBgateway  [ serialPort="COM3" ] {
...
+String iCU_3550_remote    "Remote Control"    (gCentralUnit) { channel="openwebnet:bus_thermo_cu:mybridge:CU_3550:remoteControl" }
+String iCU_3550_battery   "Battery Status"    (gCentralUnit) { channel="openwebnet:bus_thermo_cu:mybridge:CU_3550:batteryStatus" }
+String iCU_3550_mode      "Mode"              (gCentralUnit) { channel="openwebnet:bus_thermo_cu:mybridge:CU_3550:mode" }
+String iCU_3550_wpn       "Weekly Program"    (gCentralUnit) { channel="openwebnet:bus_thermo_cu:mybridge:CU_3550:weeklyProgram" } 
+String iCU_3550_spn       "Scenario Program" (gCentralUnit) { channel="openwebnet:bus_thermo_cu:mybridge:CU_3550:scenarioProgram" } 
 
 Group   gLivingRoomZone                         "Living Room Zone"   { ga="Thermostat" [ modes="auto=GENERIC,heat=HEATING,cool=COOLING", thermostatTemperatureRange="7,35", useFahrenheit=false ] }
 Number:Temperature  iLR_zone_temp               "Temperature [%.1f %unit%]"   (gLivingRoomZone) { channel="openwebnet:bus_thermo_zone:mybridge:LR_zone:temperature", ga="thermostatTemperatureAmbient" }
```
Comment (lolodomo) [inline bundles/org.openhab.binding.openwebnet/README.md:301]: You added the file dashboard-tz-standalone.png but it is not referenced in README.

**Review by lolodomo — CHANGES_REQUESTED** [review]

**Comment by lolodomo** [comment]
@aconte80 : I merged another PR for this binding which introduced file conflicts in your PR. Can you please fix them ?

**Comment by lolodomo** [comment]
@mvalla: would you like to have a look to this PR ?

File: bundles/org.openhab.binding.openwebnet/README.md [hunk 6]
```diff
@@ -173,13 +176,19 @@ Systems with Central Units (4 or 99 zones) are not fully supported yet.
 | Channel Type ID (channel ID) | Applies to Thing Type IDs           | Item Type          | Description                                       | Read/Write | Advanced |
 | ---------------------------- | ----------------------------------- | ------------------ | ------------------------------------------------- | :--------: | :------: |
 | `temperature`                | `bus_thermo_zone`, `bus_thermo_sensor` | Number:Temperature | The zone currently sensed temperature       | R          | N        |
-| `setpointTemperature`        | `bus_thermo_zone`                    | Number:Temperature | The zone setpoint temperature           | R/W        | N        |
+| `setpointTemperature`        | `bus_thermo_zone`, `bus_thermo_cu`                | Number:Temperature | The zone setpoint temperature           | R/W        | N        |
```
[the remark this answers is not in the document]
Comment (aconte80) [reply bundles/org.openhab.binding.openwebnet/README.md:179]: fixed all grids alignment

File: bundles/org.openhab.binding.openwebnet/README.md (hunk 4 above)
Comment (aconte80) [reply bundles/org.openhab.binding.openwebnet/README.md:298]: fixed

File: bundles/org.openhab.binding.openwebnet/README.md (hunk 3 above)
Comment (aconte80) [reply bundles/org.openhab.binding.openwebnet/README.md:190]: fixed

File: bundles/org.openhab.binding.openwebnet/README.md (hunk 5 above)
Comment (aconte80) [reply bundles/org.openhab.binding.openwebnet/README.md:301]: removed. I've used it in a community post , so wrong place for saving it!

File: bundles/org.openhab.binding.openwebnet/src/main/java/org/openhab/binding/openwebnet/internal/handler/OpenWebNetThermoregulationHandler.java [hunk 7]
```diff
@@ -177,6 +247,13 @@ private void handleMode(Command command) {
         }
     }
 
+    private String getWhere(String where) {
+        if (isCentralUnit)
+            return "#0";
+        else
+            return isStandAlone ? where : "#" + where;
```
Comment (lolodomo) [inline bundles/org.openhab.binding.openwebnet/src/main/java/org/openhab/binding/openwebnet/internal/handler/OpenWebNetThermoregulationHandler.java:254]: Please use `{ }`

File: bundles/org.openhab.binding.openwebnet/src/main/java/org/openhab/binding/openwebnet/internal/handler/OpenWebNetThermoregulationHandler.java [hunk 8]
```diff
@@ -199,6 +276,27 @@ private void handleFunction(Command command) {
     @Override
     protected void handleMessage(BaseOpenMessage msg) {
         super.handleMessage(msg);
+
+        if (isCentralUnit) {
+            // 26/2 TODO
+            // WEEKLY torna MANUAL se le seguenti righe sono attive
+            // se le commento funziona bene ma non vengono recepite le impostazini da CU
+            // se metto SETTIMANALE 2 su CU => arriva WEEKLY_2 sulla combo mode che ovviamente non esiste e quindi si torna a manual
```
Comment (lolodomo) [inline bundles/org.openhab.binding.openwebnet/src/main/java/org/openhab/binding/openwebnet/internal/handler/OpenWebNetThermoregulationHandler.java:284]: Please provide an English comment

**Review by lolodomo — CHANGES_REQUESTED** [review]

File: bundles/org.openhab.binding.openwebnet/src/main/java/org/openhab/binding/openwebnet/internal/handler/OpenWebNetThermoregulationHandler.java (hunk 8 above)
Comment (aconte80) [reply bundles/org.openhab.binding.openwebnet/src/main/java/org/openhab/binding/openwebnet/internal/handler/OpenWebNetThermoregulationHandler.java:284]: my bad, removed

File: bundles/org.openhab.binding.openwebnet/src/main/java/org/openhab/binding/openwebnet/internal/handler/OpenWebNetThermoregulationHandler.java (hunk 7 above)
Comment (aconte80) [reply bundles/org.openhab.binding.openwebnet/src/main/java/org/openhab/binding/openwebnet/internal/handler/OpenWebNetThermoregulationHandler.java:254]: done

**Comment by aconte80** [comment]
thanks @lolodomo for the review, just fixed the conflicts

File: bundles/org.openhab.binding.openwebnet/README.md [hunk 9]
```diff
@@ -150,36 +150,45 @@ Temperature sensors can be configured defining a `bus_thermo_sensor` Thing with
...
+| ---------------------------- | -------------------------------------- | ------------------ | --------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------- | :--------: | :------: |
+| `temperature`                | `bus_thermo_zone`, `bus_thermo_sensor` | Number:Temperature | The zone currently sensed temperature                                                                                                                                                   | R          | N        |
+| `setpointTemperature`        | `bus_thermo_zone`, `bus_thermo_cu`     | Number:Temperature | The zone setpoint temperature                                                                                                                                                           | R/W        | N        |
```
Comment (mvalla) [inline bundles/org.openhab.binding.openwebnet/README.md:179]:
```suggestion
| `setpointTemperature`        | `bus_thermo_zone`, `bus_thermo_cu`     | Number:Temperature | The zone or Central Unit setpoint temperature                                                                                                                                                           | R/W        | N        |
```

File: bundles/org.openhab.binding.openwebnet/README.md [hunk 10]
```diff
@@ -218,6 +227,15 @@ See [openwebnet.sitemap](#openwebnet-sitemap) & [openwebnet.rules](#openwebnet-r
         - `EXTENDED_PRESS` - sent after `START_EXTENDED_PRESS` if you keep the button pressed longer; will be sent again every 0,5sec as long as you hold pressed (good for dimming rules)
         - `RELEASE_EXTENDED_PRESS` - sent once when you finally release the button after having it pressed longer than 0,5sec
 
+####  `mode` for values WEEKLY and SCENARIO (Central Unit)
+
+There are three WEEKLY and sixteen SCENARIO programs defined in the central unit.
```
Comment (mvalla) [inline bundles/org.openhab.binding.openwebnet/README.md:232]:
```suggestion
There are three WEEKLY and sixteen SCENARIO programs defined for the Central Unit.
```

File: bundles/org.openhab.binding.openwebnet/src/main/resources/OH-INF/thing/BusThermoCentralUnit.xml [hunk 11]
```diff
@@ -0,0 +1,43 @@
...
+			<property name="model">BTicino Central Unit 3550 (99 zones) or BTI-L/NT/HC/4695 (4 zones) </property>
+			<property name="ownDeviceType">430</property>
+		</properties>
+
+		<representation-property>ownId</representation-property>
+
+		<config-description>
+			<parameter name="where" type="text" readOnly="true">
+				<label>OpenWebNet Address (where)</label>
+				<description>The Central Unit can only assume where=0. </description>
```
Comment (mvalla) [inline bundles/org.openhab.binding.openwebnet/src/main/resources/OH-INF/thing/BusThermoCentralUnit.xml:38]:
```suggestion
				<description>The Central Unit can only assume where=0.</description>
```

File: bundles/org.openhab.binding.openwebnet/src/main/resources/OH-INF/thing/BusThermoCentralUnit.xml [hunk 12]
```diff
@@ -0,0 +1,43 @@
...
+	xsi:schemaLocation="https://openhab.org/schemas/thing-description/v1.0.0 https://openhab.org/schemas/thing-description-1.0.0.xsd">
+
+	<!-- Thing for BUS Thermo Central Unit -->
+	<thing-type id="bus_thermo_cu">
+		<supported-bridge-type-refs>
+			<bridge-type-ref id="bus_gateway"/>
+		</supported-bridge-type-refs>
+
+		<label>Thermo Central Unit</label>
+		<description>A OpenWebNet BUS/SCS configured thermo central unit.</description>
```
Comment (mvalla) [inline bundles/org.openhab.binding.openwebnet/src/main/resources/OH-INF/thing/BusThermoCentralUnit.xml:14]:
```suggestion
		<description>A OpenWebNet BUS/SCS configured thermo Central Unit.</description>
```

File: bundles/org.openhab.binding.openwebnet/src/main/resources/OH-INF/thing/channels.xml [hunk 13]
```diff
@@ -171,6 +171,103 @@
...
+				<option value="MINUS_2">-2</option>
+				<option value="MINUS_3">-3</option>
+			</options>
+		</state>
+	</channel-type>
+
+	<channel-type id="modeCentralUnit">
+		<item-type>String</item-type>
+		<label>Central Unit Mode</label>
+		<description>Set mode of the central unit (read/write)</description>
```
Comment (mvalla) [inline bundles/org.openhab.binding.openwebnet/src/main/resources/OH-INF/thing/channels.xml:196]:
```suggestion
		<description>Set mode of the Central Unit (read/write)</description>
```

File: bundles/org.openhab.binding.openwebnet/src/main/resources/OH-INF/thing/channels.xml [hunk 14]
```diff
@@ -171,6 +171,103 @@
...
+				<option value="WEEKLY">Weekly</option>
+				<option value="SCENARIO">Scenario</option>
+			</options>
+		</state>
+	</channel-type>
+
+	<channel-type id="weeklyProgramCentralUnit">
+		<item-type>Number</item-type>
+		<label>Weekly Program Number</label>
+		<description>Set program number in the central unit, valid only for Mode Central Unit = "WEEKLY"(read/write)</description>
```
Comment (mvalla) [inline bundles/org.openhab.binding.openwebnet/src/main/resources/OH-INF/thing/channels.xml:211]:
```suggestion
		<description>Set weekly program number for the Central Unit, valid only with Central Unit mode = "WEEKLY" (read/write)</description>
```

File: bundles/org.openhab.binding.openwebnet/src/main/resources/OH-INF/thing/channels.xml [hunk 15]
```diff
@@ -171,6 +171,103 @@
...
+				<option value="2">Program 2</option>
+				<option value="3">Program 3</option>
+			</options>
+		</state>
+	</channel-type>
+
+	<channel-type id="scenarioProgramCentralUnit">
+		<item-type>Number</item-type>
+		<label>Scenario Program Number</label>
+		<description>Set program number in the central unit, valid only for Mode Central Unit = "SCENARIO"(read/write)</description>
```
Comment (mvalla) [inline bundles/org.openhab.binding.openwebnet/src/main/resources/OH-INF/thing/channels.xml:224]:
```suggestion
		<description>Set scenario program number for the Central Unit, valid only with Central Unit mode = "SCENARIO" (read/write)</description>
```

File: bundles/org.openhab.binding.openwebnet/src/main/java/org/openhab/binding/openwebnet/internal/handler/OpenWebNetThermoregulationHandler.java [hunk 16]
```diff
@@ -156,6 +182,21 @@ private void handleSetFanSpeed(Command command) {
         }
     }
 
+    private void handleSetProgramNumber(Command command) {
+        if (command instanceof DecimalType) {
+            if (!isCentralUnit) {
+                logger.warn("handleSetProgramNumber() This command can be sent only to a Central Unit. ");
```
Comment (mvalla) [inline bundles/org.openhab.binding.openwebnet/src/main/java/org/openhab/binding/openwebnet/internal/handler/OpenWebNetThermoregulationHandler.java:188]:
```suggestion
                logger.warn("handleSetProgramNumber() This command can be sent only for a Central Unit.");
```

File: bundles/org.openhab.binding.openwebnet/src/main/java/org/openhab/binding/openwebnet/internal/handler/OpenWebNetThermoregulationHandler.java [hunk 17]
```diff
@@ -336,4 +405,82 @@ private void updateActuatorStatus(Thermoregulation tmsg) {
...
+        logger.debug("updateCUBatteryStatus(): {}", status);
+    }
+
+    private Boolean channelExists(String channelID) {
+        return thing.getChannel("openwebnet:" + channelID) != null;
+    }
+
+    @Override
+    protected void refreshDevice(boolean refreshAll) {
+        if (isCentralUnit) {
```
Comment (mvalla) [inline bundles/org.openhab.binding.openwebnet/src/main/java/org/openhab/binding/openwebnet/internal/handler/OpenWebNetThermoregulationHandler.java:437]:
```suggestion
        logger.debug("--- refreshDevice() : refreshing SINGLE... ({})", thing.getUID());
        if (isCentralUnit) {
```

File: bundles/org.openhab.binding.openwebnet/src/main/java/org/openhab/binding/openwebnet/internal/handler/OpenWebNetThermoregulationHandler.java [hunk 18]
```diff
@@ -336,4 +405,82 @@ private void updateActuatorStatus(Thermoregulation tmsg) {
...
+                send(Thermoregulation.requestStatus("#0"));
+            } catch (OWNException e) {
+                logger.warn("refreshDevice() central unit returned OWNException {}", e.getMessage());
+            }
+
+            return;
+        }
+
+        if (deviceWhere != null) {
+            logger.debug("--- refreshDevice() : refreshing SINGLE... ({})", thing.getUID());
```
Comment (mvalla) [inline bundles/org.openhab.binding.openwebnet/src/main/java/org/openhab/binding/openwebnet/internal/handler/OpenWebNetThermoregulationHandler.java:453]:
```suggestion
```

File: bundles/org.openhab.binding.openwebnet/src/main/java/org/openhab/binding/openwebnet/internal/discovery/OpenWebNetDeviceDiscoveryService.java [hunk 19]
```diff
@@ -144,7 +144,10 @@ public void newDiscoveryResult(Where where, OpenDeviceType deviceType, @Nullable
                 break;
             }
             case SCS_THERMO_CENTRAL_UNIT: {
-                logger.warn("newDiscoveryResult() deviceType={} is not supported yet (WHERE={})", deviceType, where);
+                // TODO: is this required? no way to enter here...
```
Comment (mvalla) [inline bundles/org.openhab.binding.openwebnet/src/main/java/org/openhab/binding/openwebnet/internal/discovery/OpenWebNetDeviceDiscoveryService.java:147]:
```suggestion
```

**Review by mvalla — CHANGES_REQUESTED** [review]
@lolodomo @aconte80: I completed a review with only minor changes. LGTM.

**Comment by lolodomo** [comment]
@mvalla : all your comments were considered. So I merge. Thank you @aconte80

**Pull request merged**